state/application.go

```diff
@@ -832,79 +897,61 @@ func (s *Application) SetCharm(cfg SetCharmConfig) error {
 		}
 	}
 
-	services, closer := s.st.getCollection(applicationsC)
-	defer closer()
-
-	// this value holds the *previous* charm modified version, before this
-	// transaction commits.
-	var charmModifiedVersion int
+	var newCharmModifiedVersion int
 	channel := string(cfg.Channel)
+	scopy := &Application{s.st, s.doc}
 	buildTxn := func(attempt int) ([]txn.Op, error) {
+		s := scopy
 		if attempt > 0 {
-			// NOTE: We're explicitly allowing SetCharm to succeed
-			// when the application is Dying, because service/charm
-			// upgrades should still be allowed to apply to dying
-			// services and units, so that bugs in departed/broken
-			// hooks can be addressed at runtime.
-			if notDead, err := isNotDeadWithSession(services, s.doc.DocID); err != nil {
+			if err := s.Refresh(); err != nil {
 				return nil, errors.Trace(err)
-			} else if !notDead {
-				return nil, ErrDead
 			}
 		}
 
-		// We can't update the in-memory application doc inside the transaction, so
-		// we manually udpate it at the end of the SetCharm method. However, we
-		// have no way of knowing what the charmModifiedVersion will be, since
-		// it's just incrementing the value in the DB (and that might be out of
-		// step with the value we have in memory).  What we have to do is read
-		// the DB, store the charmModifiedVersion we get, run the transaction,
-		// assert in the transaction that the charmModifiedVersion hasn't
-		// changed since we retrieved it, and then we know what its value must
-		// be after this transaction ends.  It's hacky, but there's no real
-		// other way to do it, thanks to the way mgo's transactions work.
-		var doc applicationDoc
-		err := services.FindId(s.doc.DocID).One(&doc)
-		var charmModifiedVersion int
-		switch {
-		case err == mgo.ErrNotFound:
-			// 0 is correct, since no previous charm existed.
-		case err != nil:
-			return nil, errors.Annotate(err, "can't open previous copy of charm")
-		default:
-			charmModifiedVersion = doc.CharmModifiedVersion
+		// NOTE: We're explicitly allowing SetCharm to succeed
+		// when the application is Dying, because service/charm
+		// upgrades should still be allowed to apply to dying
+		// services and units, so that bugs in departed/broken
+		// hooks can be addressed at runtime.
+		if s.Life() == Dead {
+			return nil, ErrDead
 		}
+
+		// Record the current value of charmModifiedVersion, so we can
+		// set the value on the method receiver's in-memory document
+		// structure.
+		newCharmModifiedVersion = s.doc.CharmModifiedVersion
```

Comment:
We only increment the version when the charm URL changes (see the "newCharmModifiedVersion++" below). I'll add a comment to that effect.